[dynamo] Guard on nn.Module dicts and type

colesbury opened the pull request:
(WIP)

The NN_MODULE guard now subsumes guards on Module attributes. The check_fn will fail if the module attributes are changed (such as Module.training), parameters, submodules, and buffers are added or removed, and if fields are changed on the type itself.

This gives up specificity in the guard check -- if any field is changed the check_fn fails -- for faster overall checks.


cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov @soumith @desertfire

colesbury committed:
[dynamo] Guard on nn.Module dicts and type

The NN_MODULE guard now subsumes guards on Module attributes. The
check_fn will fail if the module attributes are changed (such as
Module.training), parameters, submodules, and buffers are added or
removed, and if fields are changed on the type itself.

This gives up specificity in the guard check -- if any field is changed
the check_fn fails -- for faster overall checks.

colesbury committed:
Use python_compat.h

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -464,10 +465,125 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
+
+  // NOTE: we must check the dict version tag before we check the attributes,
+  // because the attributes may be dead references if the dict has been updated.
+  PyObject* dict = PyObject_GenericGetDict(mod, NULL);
```
Trying to understand. This will basically get `mod.__dict__` and will cover changes to member variables like `training`, right?

wconstab replied:
i think this checks the version of mod.__dict__ in case, for instance, someone deleted `_parameters` from it and later on we would attempt to get the version of `mod.__dict__['_parameters']`

but, this check should also cover keys like `training` which being plain types would require changing the mod.__dict__ version if updating them?

(also just trying to check my understanding)

colesbury replied:
Yes, any changes to attributes stored on the `__dict__` will lead to a change in `PyDictObject.ma_version_tag`.

For Python 3.12+, we'll want to use a different mechanism ([dictionary watchers](https://docs.python.org/3.12/c-api/dict.html#c.PyDict_AddWatcher)) because `ma_version_tag` is deprecated.

colesbury replied:
@wconstab, yes I think your understanding is correct. If someone changes, something like `self.foo.bar = 3` that *won't* be caught it's mutating "foo" and not the module's `__dict__`.

Modifications to `_parameters`, `_buffers`, and hooks are handled in here: https://github.com/pytorch/pytorch/pull/100642/files#diff-f441b0c110772dd661a6b6cc9402f685eed03f0319597d91b661bf0bccd2d11aR473-R479.

anijain2305 replied:
>  If someone changes, something like self.foo.bar = 3 that won't be caught it's mutating "foo" and not the module's __dict__

What about submodules? Like suppose we have larger nn.Module instance which instantiates many instances of smaller nn.Module instances in its `__init__` function. Do we need `NNModuleGuard` for each of these submodules?

wconstab replied:
i think we'd need a guard on each nnmodule.  the handling on the root module's `_modules` dict would help notice any newly added or deleted submodule, but i don't think it would notice if anything inside that submodule got poked.

anijain2305 commented:
This looks really really interesting.

I have one example of a common but very silly graph break for nn.Modules - https://gist.github.com/anijain2305/ee21192fac2e1d8164e5888aa5f7c0fb

Here, the `SubModule` has a graph break in `forward` and we end having `__check_id(self, xxxx)` on nn.Module instances. I am not sure if this PR resolves that. Wondering it it already does that. (cc @voznesenskym )

cc @wconstab for visibility, this will resolve the NN module hook problem as well.

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -464,10 +465,125 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
+typedef struct {
+  PyObject_HEAD
+  PyObject *wr;
+  unsigned int version_tag;
+  uint64_t dict_version_tag;
```
Platform dependent and fixed sizes in the same struct?

colesbury replied:
Those are the types of CPython's `tp_version_tag` and `ma_version_tag`. I don't make the rules!

https://github.com/python/cpython/blob/v3.10.11/Include/cpython/object.h#L268-L269

https://github.com/python/cpython/blob/v3.10.11/Include/cpython/dictobject.h#L16-L18

voznesenskym commented:
Tests?

colesbury committed:
Try moving assignment to name earlier.

The assignment invalidates the nn.Module dictionary guards, which seems
to lead to perf regressions on hf_Bert.

colesbury committed:
Apply lintrunner

colesbury committed:
Don't install guard if the Module is missing attributes

colesbury committed:
Guards now catch changes to nn.Module hooks by default

colesbury commented:
https://github.com/pytorch/pytorch/actions/runs/4982793021

colesbury committed:
Temporarily disable tp_version_tag check

colesbury commented:
Run with no tp_version_tag check:
https://github.com/pytorch/pytorch/actions/runs/5007929902

anijain2305 commented:
It is possible that CSE optimization for guards affects this PR - https://github.com/pytorch/pytorch/pull/98488

Just sharing to ensure that perf dashboard runs are looked at with this knowledge.

colesbury committed:
Include identity check as part of nn.Module guard

This is necessary for models like hf_BigBird, hf_Reformer, and yolov3.

colesbury commented:
With fixed identity check for nn.Module:
https://github.com/pytorch/pytorch/actions/runs/5050043845

colesbury committed:
Update inductor expected graph breaks

colesbury committed:
The unimplemented check appears necessary for test_lazy_module1

colesbury committed:
Fix test_nn_moduledict_contains

colesbury committed:
Fix lint

colesbury committed:
Update inductor expected graph breaks (again)

colesbury committed:
Add guards for "self" if it's an nn.Module

colesbury committed:
Update nn-module.rst and check backward hooks

colesbury committed:
Add additional test case for hooks and recompilation

colesbury committed:
Update graph breaks

anijain2305 commented:
This is merged in https://github.com/pytorch/pytorch/pull/108528

eellison reviewed (commented):
Is there a way to turn this off ?
